[ONNX] Provide argument in ONNX export to exclude intializers from graph inputs.

spandantiwari opened the pull request:
Starting ONNX IR version 4, the initializers in the ONNX graph do not have to be inputs of the graphs. This constraint, which existed in IR version 3 and earlier, was relaxed in IR version 4. This PR provides an API level argument to allow ONNX export with the relaxed constraint of IR version 4, i.e. provides the option to not include initializers as inputs. This allows backends/runtimes to do certain optimizations, such as constant folding, better.

*Edit*: After discussion with @houseroad we have the following behavior. For any OperatorExportType, except OperatorExportTypes.ONNX, the current status of export is maintained in this PR by default. However, the user can override it by setting the `keep_initializers_as_inputs` argument to the export API.  But when exporting to ONNX, i.e. OperatorExportType is OperatorExportTypes.ONNX, the current status is changed in that by default the initializers are NOT part of the input. Again, the default can be overridden by setting the `keep_initializers_as_inputs` argument.

Commit:
Provide argument in ONNX export to exclude intializers from graph inputs.

Commit:
Fixed lint issues.

spandantiwari commented:
@houseroad - rebased.

spandantiwari commented:
Hi @houseroad , @bddppq - this one is all green now. Design is in line with earlier discussion with @houseroad.

bddppq reviewed (dismissed):
Looks reasonable to me

spandantiwari commented:
@bddppq - I rebased again to resolve merge conflicts.

Commit:
Fix lint and typo.

spandantiwari commented:
@bddppq - just checking to see if this is ready for merge. Thanks.

bddppq commented:
@spandantiwari Land failed because there are some tests failed. From looking at the error message it should be because caffe2 onnx backend doesn't support onnx models without initializers being part of the graph input yet (though that should be fine as long as the default behavior is unchanged, and I couldn't spot any obvious places in this diff that changes the default)

spandantiwari commented:
@bddppq - Ah, this could be due to the misleading comment in the description above. That description is outdated, and I should have updated it earlier. But I have updated it now. I had a discussion with @houseroad and we had decided to change the default just for the case when we are exporting to ONNX and keep it the same for other (ATEN_FALLBACK etc). So the behavior is that if the `keep_initializers_as_inputs` argument is specified then we just use the specified behavior. But if it not specified (default for this argument is `None` by design) then we choose the value for it automatically based on whether the export is ONNX or something else. If it is not ONNX, then we use the `keep_initializers_as_inputs=True` (existing) behavior. If it is ONNX export then we do the new behavior equivalent to  `keep_initializers_as_inputs=False`. 

Sorry for the confusion. I have updated the PR description above as well. 

Regarding failing tests, my guess is that we might have to set ` `keep_initializers_as_inputs=True` explicitly going into the calls for the export API, similar to what we are doing in `do_export()` method calls in `test_pytorch_onnx_caffe2.py` in this PR.

bddppq commented:
@spandantiwari Ah I see, now the test failures are explainable. I have located the places to add the explicit `keep_initializers_as_inputs` argument and indeed the tests are passing now.
However, I'm a little bit hesitated to merge this because it's a breaking change for users who use caffe2 backend (also since the default value of `operator_export_type` in `torch.onnx._export` is `OperatorExportTypes.ONNX` which means it's hard to locate all the call sites by simply grepping all the code).
Considering Lu is out and currently I don't have enough bandwidth to handle this breaking change, shall we modify this PR to not change the default behavior for now and when Lu is back you folks can discuss a plan to roll out a change of the default behavior?

Commit:
Removing change in default behavior. Maintaining existing behavior.

spandantiwari commented:
@bddppq - OK. I am hoping that we can make the change to the default. But I understand, and we will circle back on this once Lu is back.
For now I have updated the PR to keep the existing behavior. Basically, I have pulled out these lines from `_export` and `_export_to_pretty_string` methods in `utils.py`. 
```
        if keep_initializers_as_inputs is None and operator_export_type is OperatorExportTypes.ONNX:
            val_keep_init_as_ip = False
```
This should remove any test failure on caffe2 backend. If you can review that'll be great.

bddppq commented on torch/onnx/utils.py:
```diff
@@ -306,7 +314,7 @@ def _export(model, args, f, export_params=True, verbose=False, training=False,
             export_type=ExportTypes.PROTOBUF_FILE, example_outputs=None, propagate=False,
             opset_version=None, _retain_param_name=False, do_constant_folding=False,
-            strip_doc_string=True, dynamic_axes=None):
+            strip_doc_string=True, dynamic_axes=None, keep_initializers_as_inputs=None):
```
I think we can use `keep_initializers_as_inputs=True` and pass it down (instead of creating `val_keep_init_as_ip`)?

spandantiwari replied:
I was hoping to keep the code change to a minimum when bringing back the updated default. But sure, I can change the argument default to `keep_initializers_as_inputs=True`.

spandantiwari replied:
Done.

Commit:
Updating default directly.

spandantiwari commented:
@bddppq - the failure looks unrelated.

bddppq committed:
One more place to set default behavior to be the same as before
